keps/sig-network/3698-multi-network/README.md

```diff
@@ -0,0 +1,1114 @@
+the specific IP belongs to. This list will allow only 1 IP address per family
+(v4, v6) per PodNetwork. The Pod.PodStatus.PodIP behavior will not change.
+Proposed changes below:
+```go
+// IP address information for entries in the (plural) PodIPs field.
+// Each entry includes:
+//
+//      IP: An IP address allocated to the pod.
+//      PodNetworkName: Name of the PodNetwork the IP belongs to.
+//      InterfaceName: Name of the network interface inside the Pod.
+type PodIP struct {
+        // ip is an IP address (IPv4 or IPv6) assigned to the pod
+        IP string `json:"ip,omitempty" protobuf:"bytes,1,opt,name=ip"`
+
+        // PodNetworkName is name of the PodNetwork the IP belongs to
+        //
+        // +optional
+        PodNetworkName string `json:"podNetwork"`
+
+        // InterfaceName is name of the network interface used for this attachment
+        //
+        // +optional
+        InterfaceName string `json:"interfaceName",omitempty`
+}
+```
+Example:
+```yaml
+kind: Pod
+metadata:
+  name: pod1
+  namespace: default
+spec:
+[...]
+  networks:
+  - podNetwork: default
+    interfaceName: eth0
+    isDefaultGW4: true
+  - podNetwork: dataplane1
+    interfaceName: net1
+  - podNetworkAttachmentName: my-interface-fast
+    interfaceName: net2
+status:
+[...]
+  podIP: 192.168.5.54
+  podIPs:
+  - ip: 192.168.5.54
+    podNetwork: default
+    interfaceName: eth0
+  - ip: 10.0.0.20
+    podNetwork: dataplane1
+    interfaceName: net1
+  - ip: 2011::233
+    podNetwork: my-interface-fast
+    interfaceName: net2
+```
+The above status is expected to be populated by kubelet, but this can only happen
+after CRI provides support for the new Pod API. Because of that, initially
+kubelet will behave as it does today, without updating the additional fields.
+Until CRI catches up, the PodNetwork providers will be able to update that field
+on their own.
+
+#### DRA integration (alternative)
+We have discussed, as an alternative model, usage of the DRA API, and concluded
+that using it will be less clear for the user, compared to the explicit
+PodNetwork model, proposed above.
+
+### API server changes
+These are the changes for the API server covered by this design:
+* Provide validation webhook for PodNetwork or PodNetworkAttachment objects
+* Handle “default” PodNetwork deletion
+* Extend static Pod spec validation
+
+### Scheduler changes
+These are the changes we will do in Pod scheduler:
+* Provide active Pod spec validation
+
+When one of the multi-network validation fails, scheduler will follow the current
+“failure” path for Pod:
+* set PodScheduled condition (of the Pod) to False with appropriate error message
+* send Pod Event with same error message
+
+### Endpointslice controller changes
+Considering changes to Pod.PodStatus.PodIPs list, we must ensure that the
+controller is using the correct IPs when creating the endpoints. We will ensure
+that only IPs for "defautl" PodNetwork will be used.
```

Comment:
This requires a much deeper explanation as this is the core functionality of Services and in extension Ingress and Gateway API. Services are namespaced, so it seems that there can be Services that has Pods in different networks, - how are the EndpointSlice controller going to handle these Services?
- If there can be overlapping IPs, and some of the Pods in the service belong to different networks with overlapping, how this will be handled?
- Where is the ClusterIP going to be present, in all networks or only in one? which one?
- For headless services, how DNS will be used in each network? How the resolv.conf of the Pods will look like, specially on multi homed pods?

Reply:
This KEP does not introduces support for Services. Based on @danwinship feedback, we cannot reuse PodIPs field in status anyway, and have to add new field, which will not affect this controller, I have removed this section.